### DISCUSSION TITLE
Dynamic List Suggester Options From Note Contents?

### DISCUSSION THREAD

**Discussion by Graywaren** [mProjectsCode/obsidian-meta-bind-plugin#417] [Q&A]
Hello
I would like to create fields in my daily note template that allow me to add/remove options to a list property. Meta Bind makes this very easy, which is great, however I'd like to be able to allow for the list of options to dynamically grow and change over time in a way that will let all previous daily notes have access to those options, update any renamed options, and give the option to delete all of an option from the property in all notes or just delete it as an option and leave it on all notes, rather than needing to rewrite the hardcoded options in a daily notes template and have only the new notes have the new options. 

It seems I can do some of what I want in a clunky way, by maintaining a folder full of empty notes titled as the options and using
``INPUT[inlineListSuggester(optionQuery("path/to/folder"),useLinks(true)):exampleProperty]``
That lets me have a dynamic list of options that can dynamically change and delete an option without deleting it from any note property. However, I'd really rather not have a folder full of blank notes (or use a folder with notes I'm using that I may forget that it's using and move around or otherwise accidentally break ^^) and I'd rather not be using links as I'm not sure how that will interact with other things when I want to do something like make charts based on properties.

I'd ideally like to do this by having Meta Bind create the list of options from the contents of a note or in a custom setting, so that I can maintain the list of options I want to appear in an input pointed to it. I think it's likely impossible for me to figure out how to do parts where things can be renamed or mass deleted with the js things, but I haven't been able to figure out how to get it to just create a list of options via a note's content either. 

I've done a bunch of searching and mainly found other people asking similar questions and not any sort of answer. I've looked at the advanced use case of using a property to dynamically set the options, but haven't been able to figure out whether this can be adapted to pull the contents of a note.

I'm new to obsidian and js, so please go easy on me if this is something that is definitely not possible or I missed something obvious!
Thanks so much!

**Comment by ghost** [comment]
Please, I'm struggling with this too. As far as I know it's not possible to make an optionQuery that's not a file query. Is there any way to query a list in a note, for a lis suggester input for example?

**Reply by mProjectsCode** [reply]
No. You might be able to achieve what you want with JS though, see https://www.moritzjung.dev/obsidian-meta-bind-plugin-docs/guides/advancedusecases/

**Reply by Graywaren** [reply]
I did end up going with maintaining folders of notes, which became way more useful when I learned more about backlinks, javascipt, and dataviewjs. I think that you could definitely achieve having a metabind input get it's list of options from a list maintained in a separate note via javascipt. You'd need to pull in the file's contents and then split it on your separator into an array. Then have it be added to wherever necessary depending on what method you use.  I think if you install the js-engine plug-in you could have it get those options and change the front matter property and then do the advanced use case dynamic options for select as shown on the linked to page whenever you load the note.

You could also do it without the meta-bind-js-view part if you wanted by rendering the meta bind input via a dataviewjs codeblock, using a dv.span with the metabind input with the options array mapped to a compatible string and placed as the options. That one is probably more performance heavy though since dataview can update more frequently than js-engine iirc, so it depends on your needs. I could probably put together some demo code (no guarantees it'll be very good as I'm still learning) at some point if you think it'd be useful for you Euphemius

**Reply by Alecell** [reply]
Hey @Graywaren I would love to hear more about why you found that with backlinks, javascript and dataview it was more useful. Im facing the same problem here and I cant see why with these technologies it would be better, definitely wanted to understand it before create a js snippet do do the trick! hahaha

EDIT:
Alright, create the snippet was way easier than I expected, so for those who might want the functionality here we go:
```js


---
const file = app.vault.getAbstractFileByPath('DB/Planet Types.md');
const content = await app.vault.read(file);
const options = content.split('\n').map(x => `option(${x})`).join(", ");
const str = `\`INPUT[inlineSelect(${options}):selected]\``;
return engine.markdown.create(str);
```
Thats the note structure for it to work:
![image](https://github.com/user-attachments/assets/08802092-68fd-4687-aab6-7188327410a3)

On the snippet you just need to replace the path of the note (DB/Planet Types.md) to your actual path for the file.
Where we define the INPUT[inlineSelect]... you can replace to whatever properties you need that uses option, but its a pretty great start for newcomers on Obsidian and Meta Bind!

IMPORTANT: you should have a empty space before the --- otherwise it throws a error!

**Reply by Graywaren** [reply]
Hey!
Sure thing! I found having folders with notes that I use for optionQueries as backlinks to be more useful in most cases because:
- I can link to those notes elsewhere
- I can go to the note and open local graph view to see all of the connected notes
- I can write a definition or summary of the item in the note, which then allows me to hover over the link to it in the meta bind list to view a preview of it and therefore read the summary/definition without leaving my current note
- Less lag when combining with rendering through dataview in my experience so far

There's probably other things, but those are the big ones coming to mind right now. It's mainly a matter of is the option something that I want to interact with in other contexts or be able to use a hover preview for? If so, then having it be it's own note usually ends up being more useful. 

So, using your planet types as an example, with each planet type being it's own note in a folder that is being use as an option query for a metabind input I could link to the note about the Rochoso planet type to other notes about animals or plants that are found on that planet. Then using dataview's integration with inlinks and outlinks I could easily have dataview generate a list or table of all the plants and animals found on a planet and display that on the planet's note, which I could then easily reference by hovering over the link to the planet's note any place it appears. 

Another example, if I had a folder of notes of spells, I could write the definitions and stats for each spell in the note, then in a character sheet, use metabind to add the spells that a character can cast to a field, and once again, easy hover over reference to what the spell is, requires, and stats for it etc. Then you can stack it further by things like having a folder of spell components, that you add to each spell's note, and then you could make a table of all spells that use X component's note page. 

That said, your snippet is great! I think you may want to use [`vault.app.cachedRead(file)`](https://docs.obsidian.md/Reference/TypeScript+API/Vault/cachedRead) instead of `read(file)` since you don't want to actually modify that file's content. I've modified your snippet a little so that you can store other information in the note with the options (such as YAML frontmatter) and not have that also be added as an option. 

Here's the contents of my note that contains the options:
![image](https://github.com/user-attachments/assets/4cea4f4c-f386-4a55-ab32-7933bc602b56)
And the source view markdown of it that includes frontmatter properties:
```
---
created: 2025-05-29T15:41
modified: 2025-05-29T15:47
---
# Meta Bind Dynamic Options

These are my dynamic options for meta bind input

***
##### Options:
1
2
3
4
5
***

I use these options [[note1|here]], [[note2|here]], and [[note3|here]].

```

Here's the modified snippet to use that note to make the options for the metabind input:
````
```meta-bind-js-view
 ---
const file = app.vault.getAbstractFileByPath('test.md');
const content = await app.vault.cachedRead(file);
const options = content.split("***\n##### Options:\n")[1].split("\n***")[0].split("\n").map(x => `option(${x})`).join(", ");
const str = `\`INPUT[inlineSelect(${options}):selected]\``;
return engine.markdown.create(str);
```
````

Here's the input rendered:
![image](https://github.com/user-attachments/assets/f40a5f12-798e-4e74-ae71-5794def210e8)

**Comment by CorvusPenn** [comment]
Another way to do something like this if you don't want to create a note for every option, is to create the list suggester in an empty note, write out all of the options manually, and then simply embed the list suggester note wherever you need the list suggester using meta bind's embed codeblock. Then, whenever you want to add or remove something from the options list you just change the source note's list suggester, and that update will be shown everywhere it's been embedded.

I use this in my daily notes and it's worked really well as I figure out what options I actually need/find useful.
